src/sentry/tasks/unmerge.py

```diff
@@ -0,0 +1,544 @@
+                        'last_seen': last_seen,
+                        'times_seen': times_seen,
+                    },
+                )
+
+                if not created:
+                    instance.update(
+                        first_seen=first_seen,
+                        times_seen=F('times_seen') + times_seen,
+                    )
+
+
+def collect_release_data(caches, project, events):
+    results = {}
+
+    for event in events:
+        release = event.get_tag('sentry:release')
+
+        # TODO: Double check this!
+        if not release:
+            continue
+
+        # XXX: It's not really clear what the canonical source is for
+        # environment between the tag and the data attribute, but I'm going
+        # with data attribute for now. Right now it seems like they are
+        # intended to both be present and the same value, but I'm not really
+        # sure that has always been the case for existing values.
+        # NOTE: ``GroupRelease.environment`` is not nullable, but an empty
+        # string is OK.
+        environment = event.data.get('environment', '')
+
+        key = (
+            event.group_id,
+            environment,
+            caches[Release](
+                project.organization_id,
+                release,
+            ).id,
+        )
+
+        if key in results:
+            first_seen, last_seen = results[key]
+            results[key] = (event.datetime, last_seen)
+        else:
+            results[key] = (event.datetime, event.datetime)
+
+    return results
+
+
+def repair_group_release_data(caches, project, events):
+    attributes = collect_release_data(caches, project, events).items()
+    for (group_id, environment, release_id), (first_seen, last_seen) in attributes:
+        instance, created = GroupRelease.objects.get_or_create(
+            project_id=project.id,
+            group_id=group_id,
+            environment=environment,
+            release_id=release_id,
+            defaults={
+                'first_seen': first_seen,
+                'last_seen': last_seen,
+            },
+        )
+
+        if not created:
+            instance.update(first_seen=first_seen)
+
+
+def get_event_user_from_interface(value):
+    return EventUser(
+        ident=value.get('id'),
+        email=value.get('email'),
+        username=value.get('valuename'),
+        ip_address=value.get('ip_address'),
+    )
+
+
+def collect_tsdb_data(caches, project, events):
+    counters = defaultdict(
+        lambda: defaultdict(
+            lambda: defaultdict(int),
+        ),
+    )
+
+    sets = defaultdict(
+        lambda: defaultdict(
+            lambda: defaultdict(set),
+        ),
+    )
+
+    frequencies = defaultdict(
+        lambda: defaultdict(
+            lambda: defaultdict(
+                lambda: defaultdict(int),
+            ),
+        ),
+    )
+
+    for event in events:
+        counters[event.datetime][tsdb.models.group][event.group_id] += 1
```

Comment:
at first it wasn't clear why we have the `tsdb.models.group` bit in here since it's the only 2nd-level key used (same sitch for `sets` and `tsdb.models.users_affected_by_group`), but I guess since frequencies uses multiple models and `tsdb.incr` and `tsdb.record` obviously want the model ID, this is a more general/flexible pattern to separate the collection from the writing. once i got that i realized this is actually pretty nice because it lets this method worry about the data in particular and `repair_tsdb_data` just care about *doing* all the writes, agnostic to the models involved.

tldr "was almost confusing at first but no it makes perfect sense"